### PR TITLE
Script changes enabling composite R2R build with shared framework

### PR DESCRIPTION
**Pull request by trylek**
This change basically does three things:

1) Adds code to build-test.cmd/sh to Crossgen2 framework in composite mode;

2) Adds code to CLRTest.Crossgen.targets to run Crossgen2 in Helix in composite mode when told to do so via the CompositeBuildMode environment variable;

3) Implements YML fixes for composite pipeline support (sadly the jobs don't pass yet, apparently due to a codegen bug in CoreLib I'm currently investigating).

Thanks

Tomas

### CODE REVIEW

File: src/coreclr/build-test.cmd [hunk 1]
```diff
@@ -104,6 +105,7 @@ if /i "%1" == "buildagainstpackages"  (echo error: Remove /BuildAgainstPackages
 if /i "%1" == "skiprestorepackages"   (set __SkipRestorePackages=1&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 if /i "%1" == "crossgen"              (set __DoCrossgen=1&set __TestBuildMode=crossgen&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 if /i "%1" == "crossgen2"             (set __DoCrossgen2=1&set __TestBuildMode=crossgen2&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
+if /i "%1" == "composite"             (set __CompositeBuildMode=1&set __DoCrossgen2=1&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
```
Comment (nattress) [inline src/coreclr/build-test.cmd:108]: We need `__TestBuildMode` to be set to crossgen2 for the test exclusion logic.

```suggestion
if /i "%1" == "composite"             (set __CompositeBuildMode=1&set __DoCrossgen2=1&set __TestBuildMode=crossgen2&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
```

Comment (trylek) [reply src/coreclr/build-test.cmd:108]: Good point, will fix, thanks Simon!

File: src/coreclr/build-test.cmd [hunk 2]
```diff
@@ -621,18 +623,45 @@ exit /b 1
...
+    set __CompositeCommandLine=!__CompositeCommandLine! "@%__CompositeResponseFile%"
+    echo Building composite R2R framework^: !__CompositeCommandLine!
+    !__CompositeCommandLine!
+    set __FailedToPrecompile=!ERRORLEVEL!
+    copy /Y "!__CompositeOutputDir!\*.*" "!CORE_ROOT!\"
+)
+
 if !__FailedToPrecompile! NEQ 0 (
-    echo Failed assemblies:
+    @echo Failed assemblies:
```
Comment (nattress) [inline src/coreclr/build-test.cmd:664]: Curious; why did you add the @ here?

**Review by nattress — APPROVED** [review]
LGTM!

File: src/coreclr/build-test.cmd (hunk 2 above)
Comment (trylek) [reply src/coreclr/build-test.cmd:664]: I _think_ this was needed to make the text display with echo turned off but I must admit I'm not 100% sure.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets [hunk 3]
```diff
@@ -124,24 +153,55 @@ if defined RunCrossGen (
...
     )
+
+    set __ResponseFile=!scriptPath!\composite-r2r.dll.rsp
+    del /Q !__ResponseFile!
+
+    set __Command=!_DebuggerFullPath!
+    set __Command=!__Command! "!CORE_ROOT!\CoreRun.exe"
+    set __Command=!__Command! "!CORE_ROOT!\crossgen2\crossgen2.dll"
+    set __Command=!__Command! @"!__ResponseFile!"
+    set __Command=!__Command! !ExtraCrossGen2Args!
```
Comment (nattress) [inline src/coreclr/tests/src/CLRTest.CrossGen.targets:174]: Should we echo this in to the response file - it might make reproducing issues easier if it's all in a single file we can paste into VS debug pane.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets [hunk 4]
```diff
@@ -124,24 +153,55 @@ if defined RunCrossGen (
...
+
+    if defined CompositeBuildMode (
+        echo --composite>>!__ResponseFile!
+        echo -o:!scriptPath!composite-r2r.dll>>!__ResponseFile!
+        rem In composite mode, treat all dll's in the test folder as rooting inputs
+        echo !scriptPath!IL\*.dll>>!__ResponseFile!
+    ) else (
+        echo --inputbubble>>!__ResponseFile!
+        echo -o:!scriptPath!$(MSBuildProjectName).dll>>!__ResponseFile!
+        echo -r:%21scriptPath%21IL\*.dll>>!__ResponseFile!
```
Comment (nattress) [inline src/coreclr/tests/src/CLRTest.CrossGen.targets:184]: Maybe I'm parsing batch wrong, but is this missing the input test assembly in non-composite mode?

File: src/coreclr/tests/src/CLRTest.CrossGen.targets [hunk 5]
```diff
@@ -124,24 +153,55 @@ if defined RunCrossGen (
...
-            call !__Command!
-            set CrossGen2Status=!ERRORLEVEL!
-        )
-        call :ReleaseLock
-        IF NOT !CrossGen2Status!==0 (
-            ECHO Crossgen2 failed with exitcode - !CrossGen2Status!
-            Exit /b 1
+    call :TakeLock
+    set CrossGen2Status=0
+    if not exist "IL" (
```
Comment (nattress) [inline src/coreclr/tests/src/CLRTest.CrossGen.targets:158]: Thanks for the quality of life improvement here for re-running crossgen each time 👍

File: src/coreclr/tests/src/CLRTest.CrossGen.targets [hunk 6]
```diff
@@ -124,24 +153,55 @@ if defined RunCrossGen (
...
+        mkdir IL
+        if defined CompositeBuildMode (
+          copy *.dll IL\
+        ) else (
+          copy $(MSBuildProjectName).dll IL\
         )
     )
+
+    set __ResponseFile=!scriptPath!\composite-r2r.dll.rsp
```
Comment (nattress) [inline src/coreclr/tests/src/CLRTest.CrossGen.targets:167]: Name this differently for non-composite mode? Or choose a more generic name like crossgen.rsp

File: src/coreclr/build-test.cmd (hunk 1 above)
Comment (trylek) [reply src/coreclr/build-test.cmd]: Fixed in 2nd commit.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 3 above)
Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:174]: Well, sadly this is tricky for several reasons:

1) CoreRun.exe naturally cannot be part of the response file, that's the app to run.

2) I need to double-check whether the app to run (crossgen2.dll) can be specified in a response file, probably yes.

3) I'm not sure about CoreRun parser but for instance the managed command-line parser we're using in Crossgen2, R2RDump and SuperIlc has a different syntax for arguments on the command line vs. in the response file - e.g., file arguments mustn't be quoted in response files, there's just one argument per line etc. - so that emitting !ExtraCrossGen2Args! into the response file is potentially risky.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 4 above)
Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:184]: Ahh, right you are, thanks for spotting, will fix.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 6 above)
Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:167]: [facepalm] I though I already have but I made the change in a different runtime repo clone. Will fix for sure!

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 3 above)
Comment (nattress) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:174]: Sorry, I meant just the `ExtraCrossgen2Args`, so when we run crossgen2 in VS we have all the args in one place and don't have to glue extras on potentially.

Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:174]: Well, yes, I need to double-check whether I can just output the extra args to the response file - in case of the managed parser, you would need to replace spaces (outside of quoted sections) with line breaks in order to make the parser work properly.

Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:187]: Hmm, so it seems to me that corerun doesn't suppor response files at all and the managed parser in Crossgen2 has exactly the problem I described i.o.w. the extra command-line args have to be constructed based on knowledge whether they are going to be put directly into the command line or into a response file. I think it will be useful to switch even Crossgen1 to use response files and unify the construction of extra args everywhere but I'd prefer to do that as a separate follow-up change if you don't mind.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 4 above)
Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets]: Fixed in 3rd commit.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 6 above)
Comment (trylek) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets]: Fixed in 3rd commit.

File: src/coreclr/tests/src/CLRTest.CrossGen.targets (hunk 3 above)
Comment (nattress) [reply src/coreclr/tests/src/CLRTest.CrossGen.targets:187]: Totally fine with that
